src/hotspot/share/cds/heapShared.cpp

```diff
@@ -601,6 +641,104 @@ void HeapShared::set_has_native_pointers(oop src_obj) {
   info->set_has_native_pointers();
 }
 
+void HeapShared::start_finding_archivable_hidden_classes() {
+  if (!CDSConfig::is_dumping_invokedynamic()) {
+    return;
+  }
+  NoSafepointVerifier nsv;
+
+  init_seen_objects_table();
+
+  find_archivable_hidden_classes_helper(archive_subgraph_entry_fields);
+  if (CDSConfig::is_dumping_full_module_graph()) {
+    find_archivable_hidden_classes_helper(fmg_archive_subgraph_entry_fields);
+  }
+}
+
+void HeapShared::end_finding_archivable_hidden_classes() {
+  if (!CDSConfig::is_dumping_invokedynamic()) {
+    return;
+  }
+  NoSafepointVerifier nsv;
+
+  delete_seen_objects_table();
+}
+
+void HeapShared::find_archivable_hidden_classes_helper(ArchivableStaticFieldInfo fields[]) {
+  if (!CDSConfig::is_dumping_heap()) {
+    return;
+  }
+  for (int i = 0; fields[i].valid(); ) {
+    ArchivableStaticFieldInfo* info = &fields[i];
+    const char* klass_name = info->klass_name;
+    for (; fields[i].valid(); i++) {
+      ArchivableStaticFieldInfo* f = &fields[i];
+      if (f->klass_name != klass_name) {
+        break;
+      }
+
+      InstanceKlass* k = f->klass;
+      oop m = k->java_mirror();
+      oop o = m->obj_field(f->offset);
+      if (o != nullptr) {
+        find_archivable_hidden_classes_in_object(o);
+      }
+    }
+  }
+}
+
+class HeapShared::FindHiddenClassesOopClosure: public BasicOopIterateClosure {
+  GrowableArray<oop> _stack;
+  template <class T> void do_oop_work(T *p) {
+    // Recurse on a GrowableArray to avoid overflowing the C stack.
+    oop o = RawAccess<>::oop_load(p);
+    if (o != nullptr) {
+      _stack.append(o);
+    }
+  }
+
+ public:
+
+  void do_oop(narrowOop *p) { FindHiddenClassesOopClosure::do_oop_work(p); }
+  void do_oop(      oop *p) { FindHiddenClassesOopClosure::do_oop_work(p); }
+
+  FindHiddenClassesOopClosure(oop o) {
+    _stack.append(o);
+  }
+  oop pop() {
+    if (_stack.length() == 0) {
+      return nullptr;
+    } else {
+      return _stack.pop();
+    }
+  }
+};
+
+void HeapShared::find_archivable_hidden_classes_in_object(oop root) {
```

Comment:
If this function find archivable hidden classes, shouldn't there be a check for`is_hidden()` on the InstanceKlass being marked as `required`. Am I missing something?

Reply:
I added the `ik->is_hidden()` check as you suggested. I also renamed the `find_archivable_hidden_classes_xxx` functions to `find_required_hidden_classes_xxx()`. This is to avoid confusion with `HeapShared::is_archivable_hidden_klass()`, which means "it is possible to archive X", but not "it is necessary to archive X".